KAFKA-14133: Move StreamsMetricsImplTest to Mockito

clolov opened the pull request:
This pull requests migrates mocks from StreamsMetricsImplTest to Mockito.

clolov committed:
KAFKA-14133: Move StreamsMetricsImplTest to Mockito

clolov committed:
Partial: commit changes to StreamsMetricsImplTest

clolov committed:
Simplify using ArgumentCaptor and equality

clolov committed:
Checkstyle suggestions

clolov commented:
Local run of the failing tests passes:
```
> Configure project :
Starting build with version 3.7.0-SNAPSHOT (commit id 79afb498) using Gradle 8.3, Java 1.8 and Scala 2.13.12
Build properties: maxParallelForks=10, maxScalacThreads=8, maxTestRetries=0
> Task :storage:storage-api:processResources NO-SOURCE
> Task :streams:processResources UP-TO-DATE
> Task :storage:processResources UP-TO-DATE
> Task :group-coordinator:processResources UP-TO-DATE
> Task :streams:createStreamsVersionFile UP-TO-DATE
> Task :storage:createVersionFile UP-TO-DATE
> Task :group-coordinator:processTestResources UP-TO-DATE
> Task :raft:processResources UP-TO-DATE
> Task :raft:processTestResources UP-TO-DATE
> Task :raft:createVersionFile UP-TO-DATE
> Task :storage:storage-api:processTestResources NO-SOURCE
> Task :streams:processTestResources UP-TO-DATE
> Task :storage:storage-api:createVersionFile UP-TO-DATE
> Task :log4j-appender:processResources NO-SOURCE
> Task :core:processResources UP-TO-DATE
> Task :core:processTestResources UP-TO-DATE
> Task :connect:json:processResources UP-TO-DATE
> Task :metadata:processResources UP-TO-DATE
> Task :connect:test-plugins:processResources UP-TO-DATE
> Task :connect:runtime:processResources UP-TO-DATE
> Task :tools:processResources NO-SOURCE
> Task :metadata:processTestResources UP-TO-DATE
> Task :connect:api:processResources UP-TO-DATE
> Task :connect:transforms:processResources UP-TO-DATE
> Task :clients:processResources UP-TO-DATE
> Task :clients:createVersionFile UP-TO-DATE
> Task :clients:processTestResources UP-TO-DATE
> Task :tools:tools-api:processResources NO-SOURCE
> Task :server-common:processResources NO-SOURCE
> Task :tools:tools-api:createVersionFile UP-TO-DATE
> Task :server-common:processTestResources UP-TO-DATE
> Task :streams:test-utils:processResources NO-SOURCE
> Task :server-common:createVersionFile UP-TO-DATE
> Task :connect:runtime:processTestResources UP-TO-DATE
> Task :generator:compileJava UP-TO-DATE
> Task :generator:processResources NO-SOURCE
> Task :generator:classes UP-TO-DATE
> Task :generator:jar UP-TO-DATE
> Task :core:processMessages UP-TO-DATE
> Task :streams:processMessages UP-TO-DATE
> Task :clients:processMessages UP-TO-DATE
> Task :metadata:processMessages UP-TO-DATE
> Task :group-coordinator:processMessages UP-TO-DATE
> Task :raft:processMessages UP-TO-DATE
> Task :storage:processMessages UP-TO-DATE
> Task :clients:compileJava UP-TO-DATE
> Task :clients:classes UP-TO-DATE
> Task :tools:tools-api:compileJava UP-TO-DATE
> Task :tools:tools-api:classes UP-TO-DATE
> Task :clients:jar UP-TO-DATE
> Task :tools:tools-api:jar UP-TO-DATE
> Task :clients:processTestMessages UP-TO-DATE
> Task :server-common:compileJava UP-TO-DATE
> Task :streams:compileJava UP-TO-DATE
> Task :streams:classes UP-TO-DATE
> Task :streams:copyDependantLibs UP-TO-DATE
> Task :connect:api:compileJava UP-TO-DATE
> Task :server-common:classes UP-TO-DATE
> Task :log4j-appender:compileJava UP-TO-DATE
> Task :log4j-appender:classes UP-TO-DATE
> Task :storage:storage-api:compileJava UP-TO-DATE
> Task :log4j-appender:jar UP-TO-DATE
> Task :server-common:jar UP-TO-DATE
> Task :storage:storage-api:classes UP-TO-DATE
> Task :streams:jar UP-TO-DATE
> Task :storage:storage-api:jar UP-TO-DATE
> Task :raft:compileJava UP-TO-DATE
> Task :connect:json:compileJava UP-TO-DATE
> Task :raft:classes UP-TO-DATE
> Task :connect:transforms:compileJava UP-TO-DATE
> Task :connect:api:classes UP-TO-DATE
> Task :connect:json:classes UP-TO-DATE
> Task :raft:jar UP-TO-DATE
> Task :connect:transforms:classes UP-TO-DATE
> Task :streams:test-utils:compileJava UP-TO-DATE
> Task :streams:test-utils:classes UP-TO-DATE
> Task :connect:test-plugins:compileJava UP-TO-DATE
> Task :connect:test-plugins:classes UP-TO-DATE
> Task :streams:test-utils:copyDependantLibs UP-TO-DATE
> Task :connect:test-plugins:jar UP-TO-DATE
> Task :streams:test-utils:jar UP-TO-DATE
> Task :metadata:compileJava UP-TO-DATE
> Task :storage:compileJava UP-TO-DATE
> Task :metadata:classes UP-TO-DATE
> Task :storage:classes UP-TO-DATE
> Task :clients:compileTestJava UP-TO-DATE
> Task :clients:testClasses UP-TO-DATE
> Task :metadata:jar UP-TO-DATE
> Task :group-coordinator:compileJava UP-TO-DATE
> Task :core:compileJava NO-SOURCE
> Task :group-coordinator:classes UP-TO-DATE
> Task :storage:storage-api:compileTestJava UP-TO-DATE
> Task :storage:storage-api:testClasses UP-TO-DATE
> Task :server-common:compileTestJava UP-TO-DATE
> Task :group-coordinator:jar UP-TO-DATE
> Task :server-common:testClasses UP-TO-DATE
> Task :storage:jar UP-TO-DATE
> Task :group-coordinator:compileTestJava UP-TO-DATE
> Task :group-coordinator:testClasses UP-TO-DATE
> Task :connect:api:copyDependantLibs UP-TO-DATE
> Task :connect:api:jar UP-TO-DATE
> Task :raft:compileTestJava UP-TO-DATE
> Task :raft:testClasses UP-TO-DATE
> Task :connect:transforms:copyDependantLibs UP-TO-DATE
> Task :connect:json:copyDependantLibs UP-TO-DATE
> Task :connect:transforms:jar UP-TO-DATE
> Task :connect:json:jar UP-TO-DATE
> Task :metadata:compileTestJava UP-TO-DATE
> Task :metadata:testClasses UP-TO-DATE
> Task :connect:runtime:compileJava UP-TO-DATE
> Task :connect:runtime:classes UP-TO-DATE
> Task :tools:compileJava UP-TO-DATE
> Task :tools:classes UP-TO-DATE
> Task :core:compileScala UP-TO-DATE
> Task :core:classes UP-TO-DATE
> Task :core:compileTestJava NO-SOURCE
> Task :core:compileTestScala UP-TO-DATE
> Task :core:copyDependantLibs UP-TO-DATE
> Task :core:jar UP-TO-DATE
> Task :core:testClasses UP-TO-DATE
> Task :connect:runtime:compileTestJava UP-TO-DATE
> Task :streams:compileTestJava UP-TO-DATE
> Task :streams:testClasses UP-TO-DATE
> Task :connect:runtime:copyDependantLibs UP-TO-DATE
> Task :connect:runtime:jar UP-TO-DATE
> Task :connect:runtime:testClasses UP-TO-DATE
> Task :tools:copyDependantLibs UP-TO-DATE
> Task :tools:jar UP-TO-DATE
> Task :streams:unitTest
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldReturnSameMetricIfAlreadyCreated PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingNodeLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetCacheLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testThroughputMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveStateStoreLevelSensors PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddMinAndMaxMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testTotalMetricDoesntDecrease PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddSum PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentThreadIds PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAvgAndTotalMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCleanupThreadLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testLatencyMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddThreadLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddClientLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingClientLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddNewStoreLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewTaskLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddRateTotalSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewClientLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewStoreLevelSensorIfNoneExists PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAmountRateAndSum PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewNodeLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldMeasureLatency PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingStoreLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetClientLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testMultiLevelSensorRemoval PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddClientLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddThreadLevelImmutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddInvocationRateToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldThrowIfRateTotalSensorIsAddedWithOddTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testRemoveSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingThreadLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAvgAndMinAndMaxMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCreateMetricOnceDuringConcurrentMetricCreationRequest PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewCacheLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewThreadLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldUseSameStoreLevelSensorKeyWithSameSensorNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingCacheLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetStoreLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddRateTotalSensorWithCustomTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveClientLevelMetricsAndSensors PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetThreadLevelTagMap PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddTotalCountAndSumMetricsToSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddLatencyRateTotalSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddAmountRate PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetNewTopicLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotMeasureLatencyDueToRecordingLevel PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotAddStoreLevelMutableMetricIfAlreadyExists PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldProvideCorrectStrings PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldThrowIfLatencyRateTotalSensorIsAddedWithOddTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentTaskIds PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldReturnMetricsVersionCurrent PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotMeasureLatencyBecauseSensorHasNoMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldNotUseSameStoreLevelSensorKeyWithDifferentStoreNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddLatencyRateTotalSensorWithCustomTags PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingTopicLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldGetExistingTaskLevelSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldAddValue PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testRemoveNullSensor PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCreateNewStoreLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > testNullMetrics PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldUseSameStoreLevelSensorKeyWithTwoDifferentSensorNames PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldCleanupThreadLevelMutableMetric PASSED
Gradle Test Run :streams:unitTest > Gradle Test Executor 20 > StreamsMetricsImplTest > shouldRemoveThreadLevelSensors PASSED
Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
For more on this, please refer to https://docs.gradle.org/8.3/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
BUILD SUCCESSFUL in 2s
91 actionable tasks: 1 executed, 90 up-to-date
```

divijvaidya commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:
```diff
@@ -1293,43 +1204,33 @@ private void verifyMetric(final String name,
+        final Sensor sensor = mock(Sensor.class);
+        when(sensor.shouldRecord()).thenReturn(true);
+        when(sensor.hasMetrics()).thenReturn(true);
+        doNothing().when(sensor).record(endTime - startTime);
```
for my curiosity, does strict subs verify usage of stubs defined using `doNothin()` as well or just for subs defined using `when()`

clolov replied:
I will aim to double-check, but I was operating under this assumption!

clolov replied:
If I add `doNothing().when(sensor).checkQuotas();` in this test I get the following error when running the whole test suite:
```
org.mockito.exceptions.misusing.UnnecessaryStubbingException: 
Unnecessary stubbings detected in test class: StreamsMetricsImplTest
Clean & maintainable test code requires zero unnecessary code.
Following stubbings are unnecessary (click to navigate to relevant line of code):
  1. -> at org.apache.kafka.streams.processor.internals.metrics.StreamsMetricsImplTest.shouldMeasureLatency(StreamsMetricsImplTest.java:1209)
```
As such, I believe this means that the verification is carried out on these stubs as well.

divijvaidya commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/metrics/StreamsMetricsImplTest.java:
```diff
@@ -146,69 +135,26 @@ public class StreamsMetricsImplTest {
-
-    private Capture<String> addSensorsOnAllLevels(final Metrics metrics, final StreamsMetricsImpl streamsMetrics) {
-        final Capture<String> sensorKeys = newCapture(CaptureType.ALL);
+    private static boolean eqMetricConfig(final MetricConfig thisMetricConfig, final MetricConfig thatMetricConfig) {
```
This we are not printing details when assertion fails, can you please add a message to print the thatMetricConfig similar to `assertTrue(eqMetricConfig(expected, actual), "my message here")`

Will help in debugging failed tests and maintain parity with existing code. Maybe someone added such detailed logging in one place because debugging this was very hard!

clolov replied:
Okay, makes sense!

clolov commented:
I am able to reproduce the failing tests locally and I am trying to understand why the behaviour is not as expected.

clolov committed:
Fix build for Java 8 and Scala 2.12

ijuma commented on gradle/dependencies.gradle:
```diff
@@ -70,8 +70,9 @@ else
 // The default strategy (i.e. what `mockito-core` uses) changed to `inline` in 5.x because it works better with newer
 // Java versions.
 // We always use the `inline` strategy.
+// For Scala 2.12 we have to use the mockito-inline library because otherwise we cannot return mocks of final classes.
```
This comment is not accurate, `mockito-core` from version `5` does support that. The issue is that Scala 2.12 uses mockito 4. So, the additional check makes sense, but the comment doesn't.

clolov replied:
Yes, you are absolutely correct. Unfortunately, even with the additional check I still cannot quite make all the tests to pass, so I will continue battling this. But thank you for the review!

ijuma replied:
Yes, we don't run Scala 2.12 with Java 11+ in CI so it would not impact the PR builds.

clolov committed:
It appears this has fixed the matchers locally across Java and Scala versions

clolov commented:
No failures in `StreamsMetricsImplTest` in the last build! I will just make the final changes based on comments and this should be good to go.

clolov committed:
Address comments from review

clolov commented:
Heya @ijuma and @divijvaidya, none of the test failures are related to the test being changed in this pull request and the build passes! I hope I have addressed all of your comments and this is ready to merge but if not, let me know what else I can do!

divijvaidya reviewed (approved):
Thanks Christo. The changes look good to me.

divijvaidya commented:
Test failures are unrelated:
```
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.CooperativeStickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription(boolean).hasConsumerRack = false](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.clients.consumer/CooperativeStickyAssignorTest/Build___JDK_11_and_Scala_2_13___testLargeAssignmentAndGroupWithNonEqualSubscription_boolean__hasConsumerRack___false/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.clients.consumer.StickyAssignorTest.testLargeAssignmentAndGroupWithNonEqualSubscription(boolean).hasConsumerRack = false](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.clients.consumer/StickyAssignorTest/Build___JDK_11_and_Scala_2_13___testLargeAssignmentAndGroupWithNonEqualSubscription_boolean__hasConsumerRack___false/)
[Build / JDK 11 and Scala 2.13 / org.apache.kafka.controller.QuorumControllerTest.testFenceMultipleBrokers()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.controller/QuorumControllerTest/Build___JDK_11_and_Scala_2_13___testFenceMultipleBrokers__/)
[Build / JDK 17 and Scala 2.13 / org.apache.kafka.streams.integration.StoreQueryIntegrationTest.shouldQuerySpecificStalePartitionStoresMultiStreamThreads()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/StoreQueryIntegrationTest/Build___JDK_17_and_Scala_2_13___shouldQuerySpecificStalePartitionStoresMultiStreamThreads__/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ReplicaManagerTest.[1] enableRemoteStorage=true](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_8_and_Scala_2_12____1__enableRemoteStorage_true/)
[Build / JDK 8 and Scala 2.12 / kafka.server.ReplicaManagerTest.[1] enableRemoteStorage=true](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.server/ReplicaManagerTest/Build___JDK_8_and_Scala_2_12____1__enableRemoteStorage_true_2/)
[Build / JDK 8 and Scala 2.12 / kafka.zk.ZkMigrationIntegrationTest.[4] Type=ZK, MetadataVersion=3.7-IV0, Security=PLAINTEXT](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/kafka.zk/ZkMigrationIntegrationTest/Build___JDK_8_and_Scala_2_12____4__Type_ZK__MetadataVersion_3_7_IV0__Security_PLAINTEXT/)
[Build / JDK 21 and Scala 2.13 / integration.kafka.server.FetchFromFollowerIntegrationTest.testRackAwareRangeAssignor(String).quorum=zk](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/integration.kafka.server/FetchFromFollowerIntegrationTest/Build___JDK_21_and_Scala_2_13___testRackAwareRangeAssignor_String__quorum_zk/)
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.ConsistencyVectorIntegrationTest.shouldHaveSamePositionBoundActiveAndStandBy](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/ConsistencyVectorIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldHaveSamePositionBoundActiveAndStandBy/)
[Build / JDK 21 and Scala 2.13 / org.apache.kafka.streams.integration.RestoreIntegrationTest.shouldInvokeUserDefinedGlobalStateRestoreListener()](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-14623/6/testReport/junit/org.apache.kafka.streams.integration/RestoreIntegrationTest/Build___JDK_21_and_Scala_2_13___shouldInvokeUserDefinedGlobalStateRestoreListener__/)
```

divijvaidya commented:
@ijuma I am merging this in without waiting for your approval since the changes are straightforward and your comments have been addressed.  Please leave a comment if something is missing and we can follow up in a separate PR.

ijuma commented:
Sounds good, thanks.
